### PR TITLE
Fix racy use of ConcurrentHashMap

### PR DESCRIPTION
**Pull request by DaveCTurner**
ConcurrentHashMap does not always behave correctly if removing elements and
concurrently checking for its emptyiness. Work around this in PeerFinder by
protecting all usages with a mutex (there was only one usage unprotected by the
mutex anyway) and then we don't even need a ConcurrentHashMap at all.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
The racy use of `ConcurrentHashMap` is demonstrated as follows:

    public void testConcurrentMapIsEmptyOnConcurrentRemovals() throws InterruptedException {
        final Map<String, String> map = newConcurrentMap();
        final Thread[] threads = new Thread[2];
        final CyclicBarrier cyclicBarrier = new CyclicBarrier(threads.length, () -> map.put("foo", "bar"));

        for (int threadIndex = 0; threadIndex < threads.length; threadIndex++) {
            threads[threadIndex] = new Thread(() -> {
                for (int iteration = 0; iteration < 100000; iteration++) {
                    try {
                        cyclicBarrier.await(1, TimeUnit.SECONDS);
                    } catch (InterruptedException | BrokenBarrierException | TimeoutException e) {
                        throw new AssertionError("unexpected exception at " + iteration, e);
                    }
                    map.remove("foo");
                    assertTrue("expected empty at " + iteration, map.isEmpty());
                }
            }, "remover-" + threadIndex);
        }

        for (Thread thread : threads) {
            thread.start();
        }

        for (Thread thread : threads) {
            thread.join();
        }
    }

File: server/src/main/java/org/elasticsearch/discovery/PeerFinder.java [hunk 1]
```diff
@@ -256,15 +258,20 @@ private Peer createConnectingPeer(TransportAddress transportAddress) {
     private boolean handleWakeUp() {
         assert holdsLock() : "PeerFinder mutex not held";
 
-        boolean peersRemoved = false;
-
-        for (final Peer peer : peersByAddress.values()) {
-            peersRemoved = peer.handleWakeUp() || peersRemoved; // care: avoid short-circuiting, each peer needs waking up
+        final List<TransportAddress> peersAddressesToRemove = new ArrayList<>();
+        for (final Entry<TransportAddress, Peer> addressAndPeer : peersByAddress.entrySet()) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/discovery/PeerFinder.java]: maybe simpler to just write
```
final boolean peersRemoved = peersByAddress.values().removeIf(Peer::handleWakeUp);
```
and then further below just
```
return peersRemoved;
```

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/discovery/PeerFinder.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/discovery/PeerFinder.java]: Very slick. I pushed 96b02b0.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**